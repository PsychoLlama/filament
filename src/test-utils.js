import { graphql } from 'graphql';
import nock from 'nock';

import { createHueLoaders, huerl } from './context';
import * as resolvers from './resolvers';
import schema from './schema';

export const bridge = nock(huerl());

export const query = async ([request]) => {
  const response = await graphql(schema, request, resolvers, {
    hue: createHueLoaders(),
  });

  expect(response.errors).toBeUndefined();

  return response.data;
};

export const createLight = (fields = {}) => ({
  uniqueid: '75:45:12:38:00:fa:0c:26-1b',
  type: 'Extended color light',
  manufacturername: 'Philips',
  swversion: '5.23.1.13452',
  name: 'Light name',
  modelid: 'LCT001',

  state: {
    xy: [0.3144, 0.3301],
    reachable: true,
    colormode: 'xy',
    effect: 'none',
    alert: 'none',
    hue: 34076,
    sat: 251,
    bri: 254,
    on: true,
    ct: 153,
  },

  ...fields,
});

export const createGroup = (fields = {}) => ({
  name: 'Group name',
  class: 'Hallway',
  lights: [1, 2],
  recycle: true,
  type: 'Room',

  action: {
    alert: 'select',
    colormode: 'ct',
    effect: 'none',
    xy: [0.5, 0.5],
    hue: 10000,
    sat: 254,
    bri: 254,
    on: true,
    ct: 250,
  },

  /* eslint-disable camelcase */
  state: { all_on: false, any_on: true },
  /* eslint-enable camelcase */

  ...fields,
});

export const createScene = fields => ({
  owner: 'kn82nNLskEnks208Nla32LnLkndaekuDNKnsnks8',
  lastupdated: '2017-10-30T02:10:46',
  appdata: { version: 1 },
  lights: ['41', '42'],
  name: 'Fall Colors',
  recycle: false,
  locked: false,
  picture: '',
  version: 2,

  ...fields,
});

export const createScenes = fields => ({
  KseUksCEskA9Al2: createScene(),
  skeCskess872Ck4: createScene(),

  ...fields,
});

export const createSchedule = fields => ({
  created: '2017-10-31T13:45:56',
  localtime: 'W124/T07:45:00',
  description: 'MyRoutine',
  time: 'W124/T13:45:00',
  name: 'Morning Alarm',
  status: 'enabled',
  recycle: true,
  command: {
    address: '/api/t0k3n/groups/11/state',
    body: { on: true },
    method: 'PUT',
  },

  ...fields,
});

export const createSchedules = fields => ({
  1: createSchedule(),
  2: createSchedule(),

  ...fields,
});
