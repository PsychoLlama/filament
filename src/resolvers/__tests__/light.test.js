import { bridge, query, createLight } from '../../test-utils';
import toHexColorCode, { colorTypes, BLACK } from '../../utils/color';

describe('Light resolver', () => {
  let endpoint;
  afterEach(() => endpoint.done());

  it('returns light data', async () => {
    const light = createLight();
    endpoint = bridge.get('/lights/30').reply(200, light);

    const data = await query`{
      light(id: 30) { name }
    }`;

    expect(data.light).toEqual({ name: light.name });
  });

  it('formats the data correctly', async () => {
    const light = createLight();
    endpoint = bridge.get('/lights/35').reply(200, light);

    const data = await query`{
      light(id: 35) {
        name manufacturer version model type on reachable id uniqueId
      }
    }`;

    expect(data.light).toEqual(
      expect.objectContaining({
        manufacturer: light.manufacturername,
        reachable: light.state.reachable,
        uniqueId: light.uniqueid,
        version: light.swversion,
        model: light.modelid,
        on: light.state.on,
        type: light.type,
        id: '35',
      }),
    );
  });

  describe('color', () => {
    it('returns black if the light is off', async () => {
      const light = createLight();
      light.state.on = false;

      endpoint = bridge.get('/lights/25').reply(200, light);

      const result = await query`{
        light(id: 25) { color }
      }`;

      expect(result.light.color).toBe(BLACK);
    });

    it('returns a hex code', async () => {
      const light = createLight();
      light.state.colormode = colorTypes.HSB;
      light.state.hue = 64000;
      light.state.sat = 200;
      light.state.bri = 150;

      endpoint = bridge.get('/lights/25').reply(200, light);

      const result = await query`{
        light(id: 25) { color }
      }`;

      expect(result.light.color).toBe(toHexColorCode(light.state));
    });

    it('understands temperature colors', async () => {
      const light = createLight();
      light.state.colormode = colorTypes.TEMP;
      light.state.bri = 100;
      light.state.ct = 215;

      endpoint = bridge.get('/lights/30').reply(200, light);

      const result = await query`{
        light(id: 30) { color }
      }`;

      expect(result.light.color).toBe(toHexColorCode(light.state));
    });

    it('understands XY colorspace', async () => {
      const light = createLight();
      light.state.colormode = colorTypes.XY;
      light.state.xy = [0.234, 0.567];

      endpoint = bridge.get('/lights/35').reply(200, light);

      const result = await query`{
        light(id: 35) { color }
      }`;

      expect(result.light.color).toBe(toHexColorCode(light.state));
    });
  });
});
