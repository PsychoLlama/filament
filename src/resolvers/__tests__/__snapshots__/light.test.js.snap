// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Light resolver color returns a hex code 1`] = `"#962030"`;

exports[`Light resolver color understands XY colorspace 1`] = `"#fffe76"`;

exports[`Light resolver color understands temperature colors 1`] = `"#645764"`;
